presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3Client.java

```diff
@@ -0,0 +1,135 @@
+    public static final String S3_SOCKET_TIMEOUT = "presto.s3.socket-timeout";
+    public static final String S3_MAX_CONNECTIONS = "presto.s3.max-connections";
+    public static final String S3_SSL_ENABLED = "presto.s3.ssl.enabled";
+    public static final String S3_MAX_ERROR_RETRIES = "presto.s3.max-error-retries";
+    public static final String S3_USER_AGENT_PREFIX = "presto.s3.user-agent-prefix";
+
+    public static String s3UserAgentSuffix = "presto";
+    protected final AmazonS3Client s3Client;
+
+    public PrestoS3Client(Configuration config, HiveClientConfig clientConfig)
+    {
+        this.s3Client = createAmazonS3Client(config, clientConfig);
+    }
+
+    protected AmazonS3Client createAmazonS3Client(Configuration config, HiveClientConfig clientConfig)
+    {
+        HiveS3Config defaults = new HiveS3Config();
+        //AWS Client Configuration Parameters
+        String userAgentPrefix = config.get(S3_USER_AGENT_PREFIX, defaults.getS3UserAgentPrefix());
+        int maxErrorRetries = config.getInt(S3_MAX_ERROR_RETRIES, defaults.getS3MaxErrorRetries());
+        boolean sslEnabled = config.getBoolean(S3_SSL_ENABLED, defaults.isS3SslEnabled());
+        Duration connectTimeout = Duration.valueOf(config.get(S3_CONNECT_TIMEOUT, defaults.getS3ConnectTimeout().toString()));
+        Duration socketTimeout = Duration.valueOf(config.get(S3_SOCKET_TIMEOUT, defaults.getS3SocketTimeout().toString()));
+        int maxConnections = config.getInt(S3_MAX_CONNECTIONS, defaults.getS3MaxConnections());
+
+        if (clientConfig.isS3SelectEnabled()) {
+            s3UserAgentSuffix = "presto-select";
+        }
+
+        ClientConfiguration clientConfiguration = new ClientConfiguration()
+                .withMaxErrorRetry(maxErrorRetries)
+                .withProtocol(sslEnabled ? Protocol.HTTPS : Protocol.HTTP)
+                .withConnectionTimeout(toIntExact(connectTimeout.toMillis()))
+                .withSocketTimeout(toIntExact(socketTimeout.toMillis()))
+                .withMaxConnections(maxConnections)
+                .withUserAgentPrefix(userAgentPrefix)
+                .withUserAgentSuffix(s3UserAgentSuffix);
+
+        PrestoS3FileSystemStats stats = new PrestoS3FileSystemStats();
+        RequestMetricCollector metricCollector = new PrestoS3FileSystemMetricCollector(stats);
+        AWSCredentialsProvider awsCredentialsProvider = getAwsCredentialsProvider(config, defaults);
+        AmazonS3Client amazonS3Client = new AmazonS3Client(awsCredentialsProvider, clientConfiguration, metricCollector);
+        amazonS3Client.setS3ClientOptions(S3ClientOptions.builder().setPathStyleAccess(true).build());
+        return amazonS3Client;
+    }
+
+    protected AWSCredentialsProvider getAwsCredentialsProvider(Configuration conf, HiveS3Config defaults)
+    {
+        boolean useInstanceCredentials = conf.getBoolean(S3_USE_INSTANCE_CREDENTIALS, defaults.isS3UseInstanceCredentials());
```

Comment:
In PrestoS3FileSystem, first credentials are tried to obtain from `getAwsCredentials`. Can we stick to same order i.e., move line 104-107 before using Instance credential ?

Reply:
Yup, consistency is always better. Will do.